modules/ingestor/src/service/advisory/csaf/loader.rs

```diff
@@ -0,0 +1,82 @@
+use crate::service::advisory::csaf::util::resolve_purls;
+use crate::service::hashing::HashingRead;
+use crate::service::Error;
+use csaf::Csaf;
+use std::io::Read;
+use trustify_common::purl::Purl;
+use trustify_module_graph::graph::Graph;
+
+pub struct CsafLoader<'g> {
+    graph: &'g Graph,
+}
+
+impl<'g> CsafLoader<'g> {
+    pub fn new(graph: &'g Graph) -> Self {
+        Self { graph }
+    }
+
+    pub async fn load<L: Into<String>, R: Read>(
+        &self,
+        location: L,
+        document: R,
+    ) -> Result<String, Error> {
+        let mut reader = HashingRead::new(document);
+
+        let csaf: Csaf = serde_json::from_reader(&mut reader)?;
+
+        let tx = self.graph.transaction().await?;
+
+        let hashes = reader.hashes();
+        let sha256 = hex::encode(hashes.sha256.as_ref());
+
+        let advisory_id = csaf.document.tracking.id.clone();
+
+        let advisory = self
+            .graph
+            .ingest_advisory(&advisory_id, location, sha256, &tx)
+            .await?;
+
+        for vuln in csaf.vulnerabilities.iter().flatten() {
+            let id = match &vuln.cve {
+                Some(cve) => cve,
+                None => continue,
```

Comment:
I kinda hope this doesn't work like it looks like you want it to. Is it not weird that an assignment might continue the loop? Is it just me? If not, I'd prefer we filter the iterator instead.